src/garage/tf/core/parameter.py

```diff
@@ -9,7 +9,8 @@ def parameter(input_var,
               initializer=tf.zeros_initializer(),
               dtype=tf.float32,
               trainable=True,
-              name='parameter'):
+              name='parameter',
+              pixel_env=False):
```

Comment:
I don't think this is an appropriate API. it breaks several levels of abstraction. there's no way a parameter layer needs to know whether or no the environment uses pixels.

Reply:
@ryanjulian I discussed with Nisanth and realized that how batch size is obtained from the input.shape is different under different circumstances, e.g. shape[0] for pixel environment while shape[:-1] for recurrent policies. I am currently investigating if parameter class is correctly implemented for recurrent policies too as GaussianLSTMPolicyWithModel seems not working in benchmark. 

Maybe a better way is to create another `parameter_conv` class that handles pixel environments specifically? @nish21

Reply:
Yes, I believe @ahtsan covered most of it. The parameter class broadcasts a variable of shape `(L, )` to shape `(batch_dim, L)`. And the batch dimensions to pick is different based on whether a pixel based input is being used or a recurrent policy is being used. 
One other way to implement this would be for the class creating the parameter to specify which dimensions of the input is batch_dim.
So the API would then be 
```
parameter(*some_inputs, batch_dim=input_shape[:0])
and 
def parameter(...):
    ...
    broadcast_shape = tf.concat(values=[batch_dim, [length]])
```

Reply:
i think the right thing to do is figure out what is going on and file an issue.

for now you can add a parameter specifying which input dimension is the batch dimension.

for the issue
1. we should figure out if we can keep the batch dimension in the same position for all data (preferably at `shape[0]`)
1. if there are pressing performance reasons why we can't keep the batch_dim at 0, implement a consistent fix across the codebase
1. make sure LSTM is not broken (and fix it if it is, add tests to catch the bug)